File: memcontainers/sdk-js/core/src/git/connections.ts
/**
 * Connection-bound git remotes — credential and origin policy.
 *
 * Guest never sees secrets. CLI may pass public `url` and/or `connection` /
 * `agentos` ref. Credential splice is host-only on engine-emitted HTTP effects.
 *
 * Origin authorization uses `@mc/host` `originAllowed` (canonical http(s)
 * origin equality; empty origins = fail closed via `.some` on empty list).
 *
 * Dual-host semantic table (JS = reference; BEAM host opts must stay equivalent):
 *
 * | Concern                 | JS (this module + effect pump)                  | BEAM (`Git.Public` / `Git.Transport`)           |
 * |-------------------------|-------------------------------------------------|-------------------------------------------------|
 * | Credential source       | `ConnectionDefinition.auth` only                | connection catalog `:auth` (or bare host `:auth`)|
 * | Guest body secrets      | Reject (`token`/`auth`/… keys) — never splice   | Reject same keys (`:guest_secrets_forbidden`)   |
 * | Connection origins      | `connection.origins` (empty = fail closed)      | same on attach_git `connections:`               |
 * | Bare URL origins        | `allowOrigins` on the effect pump               | `:allowed_origins` (when connections empty)     |
 * | Push `approve`          | `policies` → `pushAction: "approve"`            | `policies` → `:approve`                         |
 * | Push `require_approval` | `policies` → `"require_approval"` + callback    | `policies` + `:on_push_approval`                |
 * | Push `block`            | `policies` → `"block"` (most restrictive wins)  | same (`Git push is blocked by policy`)          |
 * | Auth splice             | HTTP effect headers only                         | HTTP effect headers only                         |
 * | Auth kinds (catalog)    | none \| bearer \| header \| basic               | none \| bearer \| header \| basic (query rejected)    |
 * | Userinfo in URL         | Reject                                          | Reject                                          |
 * | Secrets in logs/resp    | `redactRemoteForLog` (kind only)                | `redact_origin` / info refs only / no tokens    |
 */

import { originAllowed } from "@mc/host";
import type {
  ConnectionAuth,
  ConnectionDefinition,
  ConnectionPolicyAction,
  ConnectionPolicyRule,
} from "../types.js";

/** Re-export host primitive so git call sites share one definition with catalog/splice. */
export { originAllowed };

// ── Types ───────────────────────────────────────────────────────────────────

export interface GitRemoteBinding {
  /** Public locator without userinfo (no secrets). */
  url: string;
  connectionRef?: string;
  origins: string[];
  auth: ConnectionAuth;
  pushAction: ConnectionPolicyAction;
}

export interface ResolveRemoteOptions {
  connections?: ConnectionDefinition[];
  policies?: ConnectionPolicyRule[];
  remoteUrls?: Record<string, string>;
  remoteConnections?: Record<string, string>;
}

// ── Guest secret keys (fail closed) — contracts/git.kdl ─────────────────────

/**
 * Guest host-call arguments must never carry credential material. Host
 * connections catalog + host opts own secrets; keys match BEAM `Git.Public`
 * (dual-host with JS `GUEST_SECRET_ARG_KEYS`).
 */
const GUEST_SECRET_ARG_KEYS = new Set([
  "auth", "authorization", "token", "password", "credential", "credentials", "secret",
]);

function stderrLine(code: string, detail?: string): string {
  return `git: ${code}${detail ? ` ${detail}` : ""}\n`;
}

/** Maximum nesting depth for secret-key scanning; exceeding it fails closed. */
const GUEST_SECRET_SCAN_MAX_DEPTH = 8;
/** Maximum object/array nodes visited; exceeding it fails closed. */
const GUEST_SECRET_SCAN_MAX_NODES = 256;
const MAX_CREDENTIAL_BYTES = 16 * 1024;
const MAX_HEADER_NAME_BYTES = 256;

function fail(code: number, stderr: string): { ok: false; code: number; stderr: string } {
  return { ok: false, code, stderr };
}

/**
 * True when guest args include a secret-bearing key (case-insensitive),
 * recursively through nested maps/objects and arrays, with explicit depth and
 * node-count bounds. Bound exceedance is fail-closed (treated as secrets).
 * Used by resolve + tests — dual-host: BEAM must not splice from guest body either.
 *
 * Does not stringify-and-regex; only object/array keys are inspected.
 */
export function guestArgsCarrySecrets(args: Record<string, unknown> | undefined): boolean {
  if (!args || typeof args !== "object" || Array.isArray(args)) return false;
  return scanGuestValueForSecrets(args, 0, { nodes: 0 }) !== "clean";
}

type GuestSecretScan = "clean" | "secret" | "exceeded";

function scanGuestValueForSecrets(
  value: unknown,
  depth: number,
  counter: { nodes: number },
): GuestSecretScan {
  if (depth > GUEST_SECRET_SCAN_MAX_DEPTH) return "exceeded";
  if (value === null || value === undefined) return "clean";
  if (typeof value !== "object") return "clean";

  if (Array.isArray(value)) {
    for (const item of value) {
      counter.nodes += 1;
      if (counter.nodes > GUEST_SECRET_SCAN_MAX_NODES) return "exceeded";
      const r = scanGuestValueForSecrets(item, depth + 1, counter);
      if (r !== "clean") return r;
    }
    return "clean";
  }

  // Plain object / map — keys only; never regex body text.
  for (const [key, child] of Object.entries(value as Record<string, unknown>)) {
    counter.nodes += 1;
    if (counter.nodes > GUEST_SECRET_SCAN_MAX_NODES) return "exceeded";
    if (GUEST_SECRET_ARG_KEYS.has(key.toLowerCase())) return "secret";
    const r = scanGuestValueForSecrets(child, depth + 1, counter);
    if (r !== "clean") return r;
  }
  return "clean";
}

// ── URL / origin helpers ────────────────────────────────────────────────────

/**
 * Git-specific public URL origin: http(s) only, no userinfo.
 * Host does not export its private `requestOrigin`; keep this thin wrapper for
 * locator validation (same rules as host parse).
 */
export function requestOrigin(value: string): string | null {
  try {
    const url = new URL(value);
    if ((url.protocol !== "http:" && url.protocol !== "https:") || url.username || url.password) {
      return null;
    }
    return url.origin;
  } catch {
    return null;
  }
}

/** Public git locator: http(s) only, no userinfo. Keeps path for repo URLs. */
export function publicRemoteUrl(url: string): string | null {
  try {
    const u = new URL(url);
    if (u.protocol !== "http:" && u.protocol !== "https:") return null;
    if (u.username || u.password) return null;
    u.hash = "";
    // Prefer href without trailing slash on bare origin; keep repo path as-is.
    return u.pathname === "/" || u.pathname === ""
      ? u.origin
      : `${u.origin}${u.pathname}${u.search}`;
  } catch {
    return null;
  }
}

// ── Resolve binding ─────────────────────────────────────────────────────────

export function resolveGitRemote(
  args: Record<string, unknown> | undefined,
  opts: ResolveRemoteOptions = {},
): { ok: true; binding: GitRemoteBinding } | { ok: false; code: number; stderr: string } {
  const a = args ?? {};
  const connections = opts.connections ?? [];
  const policies = opts.policies ?? [];

  // Guest body cannot carry auth secrets — reject, never splice from args.
  if (guestArgsCarrySecrets(a)) {
    return fail(1, stderrLine("guest_auth_secrets", "(use connection ref)"));
  }

  let url = String(a.url ?? "");
  let connectionRef =
    typeof a.connection === "string"
      ? a.connection
      : typeof a.agentos === "string"
        ? a.agentos
        : undefined;

  const remoteName = typeof a.remote === "string" ? a.remote : undefined;
  if (!url && remoteName) {
    url = opts.remoteUrls?.[remoteName] ?? "";
    if (!connectionRef) connectionRef = opts.remoteConnections?.[remoteName];
  }

  let conn: ConnectionDefinition | undefined;
  if (connectionRef) {
    conn = connections.find((c) => c.ref === connectionRef);
    if (!conn) {
      return fail(1, stderrLine("unknown_connection", connectionRef));
    }
    if (!url) {
      const fromSpec =
        conn.spec && "url" in conn.spec && typeof conn.spec.url === "string"
          ? conn.spec.url
          : conn.spec && "baseUrl" in conn.spec && typeof conn.spec.baseUrl === "string"
            ? conn.spec.baseUrl
            : undefined;
      url = fromSpec ?? "";
    }
  }

  if (!url) {
    return fail(2, "clone/fetch/push need args.url or args.connection\n");
  }

  // Reject embedded credentials in the public locator.
  if (requestOrigin(url) === null) {
    return fail(1, "git: remote url must be http(s) without embedded credentials\n");
  }

  const publicUrl = publicRemoteUrl(url);
  if (!publicUrl) {
    return fail(1, "git: bad remote url\n");
  }

  // Connection-bound: empty origins = fail closed; must match allowlist before auth.
  if (connectionRef && conn) {
    const allowed = conn.origins ?? [];
    if (!allowed.length || !originAllowed(allowed, publicUrl)) {
      // Prefix from git.kdl; optional detail after prefix is dual-host OK for connection ref.
      return fail(1, stderrLine("origin_not_allowlisted", `for connection ${conn.ref}`));
    }
  }

  const auth: unknown = conn?.auth ?? { kind: "none" };
  if (!validConnectionAuth(auth)) {
    return fail(1, stderrLine("invalid_auth"));
  }
  // Dual-host (git.kdl): query auth puts secrets in URLs — reject for remotes.
  if (auth.kind === "query") {
    return fail(1, stderrLine("query_auth_unsupported"));
  }

  const origin = requestOrigin(publicUrl)!;
  const origins = conn?.origins?.length ? [...conn.origins] : [origin];

  return {
    ok: true,
    binding: {
      url: publicUrl,
      connectionRef,
      origins,
      auth,
      pushAction: evaluatePushPolicy(connectionRef ?? "*", policies),
    },
  };
}

// ── Push policy ─────────────────────────────────────────────────────────────

export function evaluatePushPolicy(
  connectionRef: string,
  policies: ConnectionPolicyRule[],
): ConnectionPolicyAction {
  if (!policies.length) return "approve";
  let worst: ConnectionPolicyAction = "approve";
  for (const rule of policies) {
    if (!matchConnectionPattern(rule.pattern, connectionRef)) continue;
    worst = moreRestrictive(worst, rule.action);
  }
  return worst;
}

function moreRestrictive(
  a: ConnectionPolicyAction,
  b: ConnectionPolicyAction,
): ConnectionPolicyAction {
  const rank = { approve: 0, require_approval: 1, block: 2 };
  const left = rank[a];
  const right = rank[b];
  if (left === undefined || right === undefined) return "block";
  return right > left ? b : a;
}

export function matchConnectionPattern(pattern: string, ref: string): boolean {
  if (pattern === "*" || pattern === ref) return true;
  if (pattern.endsWith(".*")) {
    const prefix = pattern.slice(0, -2);
    return ref === prefix || ref.startsWith(prefix + ".");
  }
  return false;
}

// ── Credential splice (host-only) ───────────────────────────────────────────

export function spliceCredentialHeaders(
  auth: ConnectionAuth,
  headers: Record<string, string> = {},
): Record<string, string> {
  if (!validConnectionAuth(auth)) {
    throw new Error(stderrLine("invalid_auth").trim());
  }
  const out = { ...headers };
  switch (auth.kind) {
    case "none":
      return out;
    case "bearer":
      putCredentialHeader(out, "Authorization", `Bearer ${auth.token}`);
      return out;
    case "header":
      putCredentialHeader(out, auth.name, auth.value);
      return out;
    case "basic": {
      putCredentialHeader(
        out,
        "Authorization",
        `Basic ${base64Utf8(`${auth.username}:${auth.password}`)}`,
      );
      return out;
    }
    case "query":
      // Rejected for product remotes (contracts/git.kdl dual-host) — secrets must
      // not live in URLs. resolveGitRemote fails closed before dial.
      return out;
    default:
      return out;
  }
}

function validConnectionAuth(value: unknown): value is ConnectionAuth {
  if (!value || typeof value !== "object" || Array.isArray(value)) return false;
  const auth = value as Record<string, unknown>;
  switch (auth.kind) {
    case "none":
      return hasExactAuthKeys(auth, ["kind"]);
    case "bearer":
      return hasExactAuthKeys(auth, ["kind", "token"]) && validSecret(auth.token);
    case "header":
      return (
        hasExactAuthKeys(auth, ["kind", "name", "value"]) &&
        validHeaderName(auth.name) &&
        validSecret(auth.value)
      );
    case "basic":
      return (
        hasExactAuthKeys(auth, ["kind", "username", "password"]) &&
        validSecret(auth.username) &&
        validSecret(auth.password) &&
        !(auth.username as string).includes(":")
      );
    case "query":
      return (
        hasExactAuthKeys(auth, ["kind", "name", "value"]) &&
        validHeaderName(auth.name) &&
        validSecret(auth.value)
      );
    default:
      return false;
  }
}

function hasExactAuthKeys(value: Record<string, unknown>, expected: string[]): boolean {
  const keys = Object.keys(value);
  return keys.length === expected.length && expected.every((key) => keys.includes(key));
}

function validSecret(value: unknown): value is string {
  return (
    typeof value === "string" &&
    value.length > 0 &&
    new TextEncoder().encode(value).byteLength <= MAX_CREDENTIAL_BYTES &&
    !/[\u0000-\u001f\u007f]/u.test(value)
  );
}

function validHeaderName(value: unknown): value is string {
  return (
    typeof value === "string" &&
    value.length <= MAX_HEADER_NAME_BYTES &&
    /^[!#$%&'*+.^_`|~0-9A-Za-z-]+$/.test(value)
  );
}

function putCredentialHeader(headers: Record<string, string>, name: string, value: string): void {
  if (Object.keys(headers).some((existing) => existing.toLowerCase() === name.toLowerCase())) {
    throw new Error(`git: credential header conflicts with existing ${name} header`);
  }
  headers[name] = value;
}

/** Browser/worker/Node-neutral UTF-8 base64 (no Buffer dependency). */
function base64Utf8(value: string): string {
  const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  const bytes = new TextEncoder().encode(value);
  let out = "";
  for (let i = 0; i < bytes.length; i += 3) {
    const a = bytes[i]!;
    const hasB = i + 1 < bytes.length;
    const hasC = i + 2 < bytes.length;
    const b = hasB ? bytes[i + 1]! : 0;
    const c = hasC ? bytes[i + 2]! : 0;
    const n = (a << 16) | (b << 8) | c;
    out += alphabet[(n >>> 18) & 63];
    out += alphabet[(n >>> 12) & 63];
    out += hasB ? alphabet[(n >>> 6) & 63] : "=";
    out += hasC ? alphabet[n & 63] : "=";
  }
  return out;
}

/**
 * Product remotes do not splice secrets into URLs (dual-host with BEAM).
 * Returns `url` unchanged; query auth is rejected at resolve time.
 */
export function spliceCredentialUrl(url: string, _auth: ConnectionAuth): string {
  return url;
}

// ── Logging ─────────────────────────────────────────────────────────────────

export function redactRemoteForLog(binding: GitRemoteBinding): string {
  const cred = binding.auth.kind === "none" ? "anon" : binding.auth.kind;
  const ref = binding.connectionRef ? ` connection=${binding.connectionRef}` : "";
  // Never log full URL with query secrets; origin + path only.
  let loc = binding.url;
  try {
    const u = new URL(binding.url);
    loc = `${u.origin}${u.pathname}`;
  } catch {
    /* */
  }
  return `${loc}${ref} auth=${cred}`;
}

File: server/lib/agent_os/git/connections.ex
defmodule AgentOS.Git.Connections do
  @moduledoc """
  Connection-bound git remotes (SYSTEMS.md §11b / docs/connections.md).

  Host-owned catalog: guest may pass public `url` and/or `connection` /
  `agentos` ref only. Credential splice is host-only (SmartHttp headers).
  Empty connection `origins` fails closed — no credential to attacker URLs.

  Mirrors TS `sdk-js/core/src/git/connections.ts`.
  """

  alias AgentOS.Git.SmartHttp

  # Guest secret keys — prefer contracts/git.kdl via AgentOS.Contracts.Git.
  # Module attribute used at compile time; runtime MapSet from contract when available.
  @guest_secret_arg_keys AgentOS.Contracts.Git.guest_secret_arg_keys()

  # Bounded recursive scan; fail closed when either bound is exceeded.
  # Keep dual-host with JS `GUEST_SECRET_SCAN_MAX_*` in connections.ts.
  @guest_secret_scan_max_depth 8
  @guest_secret_scan_max_nodes 256
  @max_credential_bytes 16 * 1024
  @max_header_name_bytes 256

  @type auth :: map()
  @type connection :: map()
  @type policy_action :: :approve | :require_approval | :block
  @type policy_rule :: map()

  @type binding :: %{
          url: String.t(),
          connection_ref: String.t() | nil,
          origins: [String.t()],
          auth: auth(),
          push_action: policy_action()
        }

  @doc """
  Resolve guest/SDK remote `args` against host catalog opts.

  Options:
  * `:connections` — host connection list (`ref` / `auth` / `origins` / `spec?`)
  * `:policies` — push policy rules
  * `:allowed_origins` — global bare-URL allowlist (also optional intersection
    when a connection is bound and the list is non-empty)
  * `:auth` — host fallback auth for **bare** URLs only (never from guest body)
  * `:remote_urls` / `:remote_connections` — named remote maps (optional)

  Guest `auth` / `token` fields are **rejected** (fail closed), never spliced.

  Returns:

  * `{:ok, binding}` with `url`, `connection_ref`, `origins`, `auth`, `push_action`
  * `{:error, reason}` atoms/tuples understood by `Orchestrator` error map
  """
  @spec resolve_remote(map() | nil, keyword()) ::
          {:ok, binding()} | {:error, term()}
  def resolve_remote(args, opts \\ [])

  def resolve_remote(args, opts) when is_list(opts) do
    a = stringify_keys(args || %{})

    if guest_args_carry_secrets?(a) do
      {:error, :guest_secrets_forbidden}
    else
      do_resolve_remote(a, opts)
    end
  end

  defp do_resolve_remote(a, opts) do
    connections = normalize_connections(Keyword.get(opts, :connections, []))

    policies = normalize_policies(Keyword.get(opts, :policies, []))

    remote_urls = Keyword.get(opts, :remote_urls, %{}) || %{}
    remote_connections = Keyword.get(opts, :remote_connections, %{}) || %{}

    url = string_or_empty(Map.get(a, "url"))
    connection_ref = connection_ref_of(a)
    remote_name = string_or_empty(Map.get(a, "remote"))

    {url, connection_ref} =
      if url == "" and remote_name != "" do
        resolved_url =
          Map.get(remote_urls, remote_name) ||
            Map.get(remote_urls, safe_atom(remote_name)) ||
            ""

        resolved_url = if is_binary(resolved_url), do: resolved_url, else: ""

        cref =
          connection_ref ||
            Map.get(remote_connections, remote_name) ||
            Map.get(remote_connections, safe_atom(remote_name))

        cref = if is_binary(cref) and cref != "", do: cref, else: nil
        {resolved_url, cref}
      else
        {url, connection_ref}
      end

    conn =
      if is_binary(connection_ref) do
        Enum.find(connections, fn c -> c.ref == connection_ref end)
      else
        nil
      end

    cond do
      is_binary(connection_ref) and is_nil(conn) ->
        {:error, {:unknown_connection, connection_ref}}

      true ->
        url =
          if url == "" and is_map(conn) do
            url_from_spec(conn) || ""
          else
            url
          end

        cond do
          url == "" ->
            {:error, :bad_url}

          SmartHttp.request_origin(url) == :error ->
            {:error, :bad_remote_url}

          true ->
            case public_remote_url(url) do
              nil ->
                {:error, :bad_remote_url}

              public_url ->
                bind_after_url(public_url, connection_ref, conn, policies, opts)
            end
        end
    end
  end

  @doc """
  True when guest args include a secret-bearing key (case-insensitive),
  recursively through nested maps and lists, with explicit depth and node-count
  bounds. Bound exceedance is fail-closed. Dual-host with JS
  `guestArgsCarrySecrets`. Does not stringify-and-regex.
  """
  def guest_args_carry_secrets?(args) when is_map(args) do
    {status, _nodes} = scan_guest_value_for_secrets(args, 0, 0)
    status != :clean
  end

  def guest_args_carry_secrets?(_), do: false

  # Returns {status, nodes} where status is :clean | :secret | :exceeded.
  defp scan_guest_value_for_secrets(_value, depth, nodes)
       when depth > @guest_secret_scan_max_depth,
       do: {:exceeded, nodes}

  defp scan_guest_value_for_secrets(value, depth, nodes) when is_map(value) do
    Enum.reduce_while(value, {:clean, nodes}, fn {k, child}, {_status, n} ->
      n = n + 1

      cond do
        n > @guest_secret_scan_max_nodes ->
          {:halt, {:exceeded, n}}

        secret_key?(k) ->
          {:halt, {:secret, n}}

        true ->
          case scan_guest_value_for_secrets(child, depth + 1, n) do
            {:clean, n2} -> {:cont, {:clean, n2}}
            other -> {:halt, other}
          end
      end
    end)
  end

  defp scan_guest_value_for_secrets(value, depth, nodes) when is_list(value) do
    Enum.reduce_while(value, {:clean, nodes}, fn item, {_status, n} ->
      n = n + 1

      cond do
        n > @guest_secret_scan_max_nodes ->
          {:halt, {:exceeded, n}}

        true ->
          case scan_guest_value_for_secrets(item, depth + 1, n) do
            {:clean, n2} -> {:cont, {:clean, n2}}
            other -> {:halt, other}
          end
      end
    end)
  end

  defp scan_guest_value_for_secrets(_value, _depth, nodes), do: {:clean, nodes}

  defp secret_key?(k) when is_binary(k), do: String.downcase(k) in @guest_secret_arg_keys

  defp secret_key?(k) when is_atom(k),
    do: String.downcase(Atom.to_string(k)) in @guest_secret_arg_keys

  defp secret_key?(_), do: false

  defp bind_after_url(public_url, connection_ref, conn, policies, opts)
       when is_binary(connection_ref) and is_map(conn) do
    allowed = conn.origins || []

    cond do
      allowed == [] or not SmartHttp.origin_allowed?(allowed, public_url) ->
        {:error, {:origin_not_allowlisted_for_connection, conn.ref}}

      true ->
        # Optional global allowlist further intersects when set (non-empty).
        case Keyword.get(opts, :allowed_origins) do
          list when is_list(list) and list != [] ->
            if SmartHttp.origin_allowed?(list, public_url) do
              ok_binding(public_url, connection_ref, allowed, conn.auth, policies)
            else
              {:error, :origin_not_allowed}
            end

          :any ->
            ok_binding(public_url, connection_ref, allowed, conn.auth, policies)

          _ ->
            ok_binding(public_url, connection_ref, allowed, conn.auth, policies)
        end
    end
  end

  defp bind_after_url(public_url, _connection_ref, _conn, policies, opts) do
    # Bare URL: R32 — require non-empty global allowed_origins match.
    # Host `:auth` only; guest auth/token was rejected before resolution.
    case SmartHttp.ensure_url_allowed(public_url, opts) do
      {:ok, _origin} ->
        host_auth = normalize_auth(Keyword.get(opts, :auth))
        allowed = bare_origins(opts, public_url)

        ok_binding(public_url, nil, allowed, host_auth, policies)

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp bare_origins(opts, public_url) do
    case Keyword.get(opts, :allowed_origins) do
      list when is_list(list) and list != [] ->
        list

      :any ->
        case SmartHttp.request_origin(public_url) do
          {:ok, o} -> [o]
          _ -> []
        end

      _ ->
        case SmartHttp.request_origin(public_url) do
          {:ok, o} -> [o]
          _ -> []
        end
    end
  end

  defp ok_binding(url, connection_ref, origins, auth, policies) do
    ref_for_policy = connection_ref || "*"
    na = normalize_auth(auth)

    # Dual-host (contracts/git.kdl): query auth not supported for remotes.
    case na do
      :invalid ->
        {:error, :invalid_auth}

      %{kind: :query} ->
        {:error, :query_auth_unsupported}

      _ ->
        {:ok,
         %{
           url: url,
           connection_ref: connection_ref,
           origins: origins || [],
           auth: na,
           push_action: evaluate_push_policy(ref_for_policy, policies)
         }}
    end
  end

  defp connection_ref_of(a) do
    case Map.get(a, "connection") || Map.get(a, "agentos") do
      ref when is_binary(ref) and ref != "" -> ref
      _ -> nil
    end
  end

  defp url_from_spec(%{spec: %{url: u}}) when is_binary(u) and u != "", do: u
  defp url_from_spec(%{spec: %{"url" => u}}) when is_binary(u) and u != "", do: u
  defp url_from_spec(%{spec: %{baseUrl: u}}) when is_binary(u) and u != "", do: u
  defp url_from_spec(%{spec: %{"baseUrl" => u}}) when is_binary(u) and u != "", do: u
  defp url_from_spec(%{spec: %{base_url: u}}) when is_binary(u) and u != "", do: u
  defp url_from_spec(%{spec: nil}), do: nil
  defp url_from_spec(%{spec: _}), do: nil
  defp url_from_spec(_), do: nil

  @doc """
  Public git locator: http(s) only, no userinfo. Keeps path for repo URLs.
  """
  @spec public_remote_url(String.t()) :: String.t() | nil
  def public_remote_url(url) when is_binary(url) do
    case URI.parse(url) do
      %URI{scheme: scheme, host: host, userinfo: userinfo} = uri
      when scheme in ["http", "https"] and is_binary(host) and host != "" ->
        if userinfo not in [nil, ""] do
          nil
        else
          %URI{uri | userinfo: nil, fragment: nil}
          |> URI.to_string()
          |> String.trim_trailing("/")
        end

      _ ->
        nil
    end
  end

  def public_remote_url(_), do: nil

  @doc "Most-restrictive push policy for a connection ref."
  @spec evaluate_push_policy(String.t(), [policy_rule()]) :: policy_action()
  def evaluate_push_policy(connection_ref, policies)
      when is_binary(connection_ref) and is_list(policies) do
    policies
    |> Enum.reduce(:approve, fn rule, worst ->
      pattern = Map.get(rule, :pattern) || Map.get(rule, "pattern") || ""

      if match_connection_pattern?(pattern, connection_ref) do
        action = Map.get(rule, :action) || Map.get(rule, "action") || :approve
        action = normalize_action(action) || :approve
        more_restrictive(worst, action)
      else
        worst
      end
    end)
  end

  def evaluate_push_policy(_, _), do: :approve

  @doc "Glob-ish connection pattern: exact, `*`, or `prefix.*`."
  @spec match_connection_pattern?(String.t(), String.t()) :: boolean()
  def match_connection_pattern?("*", _ref), do: true
  def match_connection_pattern?(pattern, ref) when pattern == ref, do: true

  def match_connection_pattern?(pattern, ref)
      when is_binary(pattern) and is_binary(ref) do
    if String.ends_with?(pattern, ".*") do
      prefix = String.slice(pattern, 0, byte_size(pattern) - 2)
      ref == prefix or String.starts_with?(ref, prefix <> ".")
    else
      false
    end
  end

  def match_connection_pattern?(_, _), do: false

  # ── Normalization ──────────────────────────────────────────────────────────

  @doc false
  def normalize_connections(list) when is_list(list) do
    list
    |> Enum.map(&normalize_connection/1)
    |> Enum.reject(&is_nil/1)
  end

  def normalize_connections(_), do: []

  defp normalize_connection(%{ref: ref} = c) when is_binary(ref) and ref != "" do
    %{
      ref: ref,
      auth: normalize_auth(Map.get(c, :auth) || Map.get(c, "auth")),
      origins: normalize_origins(Map.get(c, :origins) || Map.get(c, "origins")),
      spec: Map.get(c, :spec) || Map.get(c, "spec")
    }
    |> drop_nil_spec()
  end

  defp normalize_connection(%{"ref" => ref} = c) when is_binary(ref) and ref != "" do
    normalize_connection(%{
      ref: ref,
      auth: Map.get(c, "auth") || Map.get(c, :auth),
      origins: Map.get(c, "origins") || Map.get(c, :origins),
      spec: Map.get(c, "spec") || Map.get(c, :spec)
    })
  end

  defp normalize_connection(_), do: nil

  defp drop_nil_spec(%{spec: nil} = c), do: Map.delete(c, :spec)
  defp drop_nil_spec(c), do: c

  defp normalize_policies(list) when is_list(list) do
    list
    |> Enum.map(&normalize_policy/1)
    |> Enum.reject(&is_nil/1)
  end

  defp normalize_policies(_), do: []

  defp normalize_policy(%{pattern: p, action: a} = r) when is_binary(p) do
    case normalize_action(a) do
      nil ->
        nil

      action ->
        base = %{pattern: p, action: action}

        case Map.get(r, :owner) || Map.get(r, "owner") do
          nil -> base
          owner -> Map.put(base, :owner, owner)
        end
    end
  end

  defp normalize_policy(%{"pattern" => p, "action" => a} = r) when is_binary(p) do
    normalize_policy(%{pattern: p, action: a, owner: Map.get(r, "owner")})
  end

  defp normalize_policy(_), do: nil

  defp normalize_action(a) when a in [:approve, :require_approval, :block], do: a
  defp normalize_action("approve"), do: :approve
  defp normalize_action("require_approval"), do: :require_approval
  defp normalize_action("block"), do: :block
  defp normalize_action(_), do: nil

  defp normalize_origins(nil), do: []
  defp normalize_origins(list) when is_list(list), do: Enum.filter(list, &is_binary/1)
  defp normalize_origins(_), do: []

  defp normalize_auth(nil), do: %{kind: :none}
  defp normalize_auth(:invalid), do: :invalid

  defp normalize_auth(auth) when is_map(auth) do
    case normalize_map_keys(auth) do
      :invalid -> :invalid
      normalized -> normalize_auth_map(normalized)
    end
  end

  defp normalize_auth(_), do: :invalid

  defp normalize_auth_map(normalized) do
    kind = parse_kind(Map.get(normalized, "kind"))

    case parse_kind(kind) do
      :invalid ->
        :invalid

      :none ->
        if exact_keys?(normalized, ["kind"]), do: %{kind: :none}, else: :invalid

      :bearer ->
        token = Map.get(normalized, "token")

        if exact_keys?(normalized, ["kind", "token"]) and valid_secret?(token),
          do: %{kind: :bearer, token: token},
          else: :invalid

      :header ->
        name = Map.get(normalized, "name")
        value = Map.get(normalized, "value")

        if exact_keys?(normalized, ["kind", "name", "value"]) and valid_header_name?(name) and
             valid_secret?(value),
           do: %{kind: :header, name: name, value: value},
           else: :invalid

      :basic ->
        u = Map.get(normalized, "username")
        p = Map.get(normalized, "password")

        if exact_keys?(normalized, ["kind", "username", "password"]) and valid_secret?(u) and
             valid_secret?(p) and not String.contains?(u, ":"),
           do: %{kind: :basic, username: u, password: p},
           else: :invalid

      :query ->
        name = Map.get(normalized, "name")
        value = Map.get(normalized, "value")

        if exact_keys?(normalized, ["kind", "name", "value"]) and valid_header_name?(name) and
             valid_secret?(value),
           do: %{kind: :query, name: name, value: value},
           else: :invalid
    end
  end

  defp normalize_map_keys(map) do
    Enum.reduce_while(map, %{}, fn
      {key, value}, acc when is_atom(key) or is_binary(key) ->
        normalized = if is_atom(key), do: Atom.to_string(key), else: key

        if Map.has_key?(acc, normalized) do
          {:halt, :invalid}
        else
          {:cont, Map.put(acc, normalized, value)}
        end

      _, _acc ->
        {:halt, :invalid}
    end)
  end

  defp exact_keys?(:invalid, _expected), do: false

  defp exact_keys?(map, expected) when is_map(map),
    do: MapSet.new(Map.keys(map)) == MapSet.new(expected)

  defp parse_kind(k) when k in [:none, :bearer, :header, :basic, :query], do: k
  defp parse_kind("none"), do: :none
  defp parse_kind("bearer"), do: :bearer
  defp parse_kind("header"), do: :header
  defp parse_kind("basic"), do: :basic
  defp parse_kind("query"), do: :query
  defp parse_kind(_), do: :invalid

  defp valid_secret?(value) when is_binary(value) do
    value != "" and byte_size(value) <= @max_credential_bytes and String.valid?(value) and
      not String.match?(value, ~r/[\x00-\x1f\x7f]/)
  end

  defp valid_secret?(_), do: false

  defp valid_header_name?(name) when is_binary(name),
    do:
      byte_size(name) <= @max_header_name_bytes and
        String.match?(name, ~r/^[!#$%&'*+\-.^_`|~0-9A-Za-z]+$/)

  defp valid_header_name?(_), do: false

  defp more_restrictive(a, b) do
    rank = %{approve: 0, require_approval: 1, block: 2}
    if Map.get(rank, b, 0) > Map.get(rank, a, 0), do: b, else: a
  end

  defp stringify_keys(map) when is_map(map) do
    Map.new(map, fn
      {k, v} when is_atom(k) -> {Atom.to_string(k), v}
      {k, v} when is_binary(k) -> {k, v}
      {k, v} -> {to_string(k), v}
    end)
  end

  defp stringify_keys(_), do: %{}

  defp string_or_empty(s) when is_binary(s), do: s
  defp string_or_empty(_), do: ""

  # Only existing atoms — never String.to_atom on untrusted remote names.
  defp safe_atom(name) when is_binary(name) do
    String.to_existing_atom(name)
  rescue
    ArgumentError -> nil
  end
end
